Adds UNP CRD and controller

trozet opened the pull request:
With this patch the architecture is reworked so that a controller is
generic and one is spawned per specified type in the config. Each
controller has different event "handler" drivers which are specific to
the type of controller being used. The only handler which exists with
this patch is UNP. This patch allows UNP to be automatically registered
as a CRD with k8s, and create/delete actions for instances of UNP will
be read and written into Nimbess etcd.

Other changes include:
 - Consolidating and cleaning up k8s deployment files
 - Modifying etcd port to be 52379 as 12379 is used by skydive

Todo:
 - Add update support in handler for UNP
 - Update UNP CRD with a more realistic version

Signed-off-by: Tim Rozet <trozet@redhat.com>

trozet committed:
Adds UNP CRD and controller

With this patch the architecture is reworked so that a controller is
generic and one is spawned per specified type in the config. Each
controller has different event "handler" drivers which are specific to
the type of controller being used. The only handler which exists with
this patch is UNP. This patch allows UNP to be automatically registered
as a CRD with k8s, and create/delete actions for instances of UNP will
be read and written into Nimbess etcd.

Other changes include:
 - Consolidating and cleaning up k8s deployment files
 - Modifying etcd port to be 52379 as 12379 is used by skydive

Todo:
 - Add update support in handler for UNP
 - Update UNP CRD with a more realistic version

Signed-off-by: Tim Rozet <trozet@redhat.com>

shague commented on deployments/stargazer-rbac-ns-account.yaml:
```diff
@@ -0,0 +1,45 @@
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: stargazer
```
Do we want to keep this as stargazer or make it nimbess? The roles are per resource type but also per app as it is now. I wasn't sure what access the other nimbess components would need or if they could just use the same.

trozet replied:
nimbess. Will make that change. For now we will mirror the file used by nimbess-agent:
https://github.com/nimbess/nimbess-agent/blob/master/k8s/nimbess-rbac-ns-account.yaml

shague commented on pkg/config/config.go:
```diff
@@ -35,14 +38,13 @@ type Config struct {
 		LogLevel:        "info",
-		Controllers:     "node",
-		NodeWorkers:     1,
-		UNPWorkers:      1,
```
So we don't want to spin up more than one worker?

trozet replied:
For now each resource being watched will get 1 thread. If that isn't enough in the future we can consider adding more workers, but for now it is OK.

shague commented on pkg/controller/stargazer.go:
```diff
@@ -0,0 +1,247 @@
+	var informer cache.SharedIndexInformer
+	var resType string
+
+	nimbessInformerFactory := unpinformer.NewSharedInformerFactory(kubeClient, time.Second*30)
```
does this return a new instance of the factory each time? It might be better to create a single factor once and then use it to spin up each controller informer. I believe I read that was the suggested model for better performance - to actually share the factory. Somehow it reduces the amount of communication.

trozet replied:
The factory var itself isn't significant, the real performance is about using a shared informer instead of a standalone one. The gain with using shared is on the backend k8s will only issue a single API call to update the cache for that resource, rather than 1 per subscriber. That is achieved by using a SharedInformerFactory here.

shague commented on pkg/controller/stargazer.go:
```diff
@@ -0,0 +1,247 @@
+	return nil
+}
+
+// HasSynced is required for the cache.Controller interface.
```
guess you copied this from the google example. I saw that also but found I didn't need to implement these.

shague commented on pkg/handlers/unp/unp.go:
```diff
@@ -0,0 +1,97 @@
+/*
```
would these handlers fit better up under controller? handler could be anything.

trozet replied:
probably. will change

shague commented on pkg/handlers/unp/unp.go:
```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2016 Skippbox, Ltd.
```
Skippbox

trozet replied:
will fix

shague commented on pkg/handlers/unp/unp.go:
```diff
@@ -0,0 +1,97 @@
+
+}
+
+// toNimbess translates a K8S UNP Config into a Nimbess Key/Value Pair to be written into ETCD
```
K8sToNimbess

trozet replied:
done

trozet committed:
Addresses shague comments

Changes-Include:
 - Moving handlers under controller
 - Minor license/doc fixes
 - Rename UNPConfig to UnifiedNetworkPolicy
 - Rework CRD to be more proper policy
 - Fix for delete operation not working

Signed-off-by: Tim Rozet <trozet@redhat.com>

trozet committed:
Minor fixes

Changes-Include:
 - Remove go mod download from dockerfile (not needed)
 - format fixes

Signed-off-by: Tim Rozet <trozet@redhat.com>
